File: Core/src/main/java/net/winternetwork/bedwars/core/listener/FlagListener.java
package net.winternetwork.bedwars.core.listener;

import net.winternetwork.bedwars.api.game.flag.Flag;
import net.winternetwork.bedwars.api.game.stage.Stage;
import net.winternetwork.bedwars.core.game.stage.StageManager;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.player.PlayerLoginEvent;

public class FlagListener implements Listener {

    private final StageManager stageManager = StageManager.getInstance();

    @EventHandler
    public void onDamage(EntityDamageByEntityEvent event) {
        final Stage stage = stageManager.getActualStage();

        for (Flag flag : stage.getFlags()) {
            if (flag.getEventClass() != event.getClass()) continue;

            flag.execute(event);
        }
    }

    @EventHandler
    public void onJoin(PlayerLoginEvent event) {
        final Stage stage = stageManager.getActualStage();

        for (Flag flag : stage.getFlags()) {
            if (flag.getEventClass() != event.getClass()) continue;

            flag.execute(event);
        }
    }

    @EventHandler
    public void onPlace(BlockPlaceEvent e) {
        final Stage stage = stageManager.getActualStage();

        for (Flag flag : stage.getFlags()) {
            if (flag.getEventClass() != BlockPlaceEvent.class) continue;

            flag.execute(e);
        }
    }

    @EventHandler
    public void onBreak(BlockBreakEvent e) {
        final Stage stage = stageManager.getActualStage();

        for (Flag flag : stage.getFlags()) {
            if (flag.getEventClass() != BlockPlaceEvent.class) continue;

            flag.execute(e);
        }
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/modules/stage/listener/FlagListener.kt
package net.winternetwork.bedwars.modules.stage.listener

import net.winternetwork.bedwars.api.game.flag.Flag
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.player.PlayerLoginEvent

class FlagListener(stageModule: _root_ide_package_.net.winternetwork.bedwars.modules.stage.StageModule) : Listener {

    private val stageManager = stageModule.stageManager

    @EventHandler
    fun onDamage(event: EntityDamageByEntityEvent) {
        val stage = stageManager?.actualStage ?: return

        val flags = stage.flags
        for (flag in flags) {
            if (flag.eventClass != event.javaClass) continue

            (flag as Flag<EntityDamageByEntityEvent>).execute(event)
        }
    }

    @EventHandler
    fun onJoin(event: PlayerLoginEvent) {
        val stage = stageManager?.actualStage ?: return

        val flags = stage.flags
        for (flag in flags) {
            if (flag.eventClass != event.javaClass) continue

            (flag as Flag<PlayerLoginEvent>).execute(event)
        }
    }

    @EventHandler
    fun onPlace(event: BlockPlaceEvent) {
        val stage = stageManager?.actualStage ?: return

        val flags = stage.flags
        for (flag in flags) {
            if (flag.eventClass != event.javaClass) continue

            (flag as Flag<BlockPlaceEvent>).execute(event)
        }
    }

    @EventHandler
    fun onBreak(event: BlockBreakEvent) {
        val stage = stageManager?.actualStage ?: return

        val flags = stage.flags
        for (flag in flags) {
            if (flag.eventClass != event.javaClass) continue

            (flag as Flag<BlockBreakEvent>).execute(event)
        }
    }
}